crates/build/src/docker.rs

```diff
@@ -0,0 +1,346 @@
+            image: Some(build_image.id.clone()),
+            entrypoint: Some(entrypoint),
+            cmd: None,
+            labels: Some(labels),
+            host_config: host_cfg,
+            attach_stderr: Some(true),
+            tty: Some(true),
+            ..Default::default()
+        };
+        let options = Some(CreateContainerOptions {
+            name: container_name.as_str(),
+            platform: Some("linux/amd64"),
+        });
+
+        let container_id: String = if containers.is_empty() {
+            client
+                .create_container(options.clone(), config.clone())
+                .await?
+                .id
+        } else {
+            let c = containers
+                .first()
+                .context("Error finding existing container")?
+                .clone();
+            if c.labels
+                .context("image does not have labels")?
+                .get("cmd_digest")
+                == Some(&digest.to_string())
+            {
+                c.id.context("Container does not have an id")?
+            } else {
+                client
+                    .create_container(options.clone(), config.clone())
+                    .await?
+                    .id
+            }
+        };
+
+        client
+            .start_container::<String>(&container_id, None)
+            .await?;
+
+        build_steps.increment_current();
+        maybe_println!(
+            verbosity,
+            " {} {}\n{}",
+            format!("{build_steps}").bold(),
+            "Started the build inside the container"
+                .bright_green()
+                .bold(),
+            "This might take a while. Check container logs for more details."
+        );
+
+        let options = Some(WaitContainerOptions {
+            condition: "not-running",
+        });
+
+        let mut wait_stream = client.wait_container(&container_id, options);
+        while wait_stream.next().await.is_some() {}
+
+        client
+            .stop_container(&container_id, Some(StopContainerOptions { t: 20 }))
+            .await?;
+
+        build_steps.increment_current();
+        maybe_println!(
+            verbosity,
+            " {} {}",
+            format!("{build_steps}").bold(),
+            "Docker container has finished the build."
+                .bright_green()
+                .bold(),
+        );
+
+        let result_contents = match std::fs::read_to_string(&file_path) {
+            Ok(content) => {
+                std::fs::remove_file(&file_path)?;
+                content
+            }
+            Err(e) => {
+                std::fs::remove_file(&file_path)?;
+                anyhow::bail!(e);
+            }
+        };
+
+        let mut build_result: BuildResult = serde_json::from_str(&result_contents)
+            .map_err(|_| {
+                anyhow::anyhow!(
+                    "Error parsing output from docker build. The build probably failed!"
+                )
+            })?;
```

Comment:
There's a shorthand for that [`serde_json::from_reader`](https://docs.rs/serde_json/latest/serde_json/fn.from_reader.html).